File: fs/src/main/scala/farjs/fs/FSFreeSpace.scala
package farjs.fs

import farjs.filelist.FileListActions.FileListDiskSpaceUpdatedAction
import farjs.filelist.api.FileListDir
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

object FSFreeSpace extends FunctionComponent[FSFreeSpaceProps] {

  private[fs] var fsService: FSService = FSService.instance

  protected def render(compProps: Props): ReactElement = {
    val props = compProps.plain
    val currDirRef = useRef[FileListDir](null)
    currDirRef.current = props.currDir

    useLayoutEffect({ () =>
      val currDir = props.currDir
      fsService.readDisk(currDir.path).toFuture.recover {
        case _ => js.undefined
      }.foreach { maybeDisk =>
        maybeDisk.toOption.foreach { disk =>
          if (isSameInstance(currDir, currDirRef.current)) {
            props.dispatch(FileListDiskSpaceUpdatedAction(disk.free))
          }
        }
      }
      ()
    }, List(System.identityHashCode(props.currDir)))

    null
  }
  
  private def isSameInstance(o1: Any, o2: Any): Boolean =
    System.identityHashCode(o1) == System.identityHashCode(o2)
}

File: fs/FSFreeSpace.mjs
/**
 * @import { Dispatch } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListDir } from "@farjs/filelist/api/FileListDir.mjs"
 * @import { FileListAction } from "@farjs/filelist/FileListActions.mjs"
 * @typedef {import("./FSService.mjs").FSService} FSService
 */
import React, { useLayoutEffect, useRef } from "react";
import FSService from "./FSService.mjs";

/**
 * @typedef {{
 *  readonly dispatch: Dispatch;
 *  readonly currDir: FileListDir;
 * }} FSFreeSpaceProps
 */

/**
 * @param {FSFreeSpaceProps} props
 */
const FSFreeSpace = (props) => {
  const { fsService } = FSFreeSpace;

  const currDirRef = /** @type {React.MutableRefObject<FileListDir>} */ (
    useRef()
  );
  currDirRef.current = props.currDir;

  useLayoutEffect(() => {
    const currDir = props.currDir;
    fsService.readDisk(currDir.path).then(
      (disk) => {
        if (disk && currDir === currDirRef.current) {
          /** @type {FileListAction} */
          const action = {
            action: "FileListDiskSpaceUpdatedAction",
            diskSpace: disk.free,
          };
          props.dispatch(action);
        }
      },
      () => undefined,
    );
  }, [props.currDir]);

  return null;
};

FSFreeSpace.displayName = "FSFreeSpace";
/** @type {FSService} */
FSFreeSpace.fsService = FSService.instance;

export default FSFreeSpace;
